### PR TITLE
Python API - delta.appendOnly enforcement

### PR DESCRIPTION
**Pull request by WarSame**
# Description
Adding enforcement of `delta.appendOnly` to the Python API. If `delta.appendOnly` is passed in as a key to the configuration and the writing `mode` is not `append` then the write will fail with a ValueError.

# Related Issue(s)
This deals with the first bullet point in #575 . The second bullet point about column invariants is not handled here, so this PR does not close the issue on its own.

# Documentation

This configuration field is already covered in the `usage.rst` -> "Examining a Table" -> "Metadata" section.

### CODE REVIEW

**Review by wjones127 — COMMENTED** [review]
Thanks for making this PR!

It looks like you've tested for the case where the configuration is passed while creating the table, but what about if the table already exists? That configuration should be saved with the table. To accomplish this, we should be able to read the existing table configuration from the metadata property: https://github.com/delta-io/delta-rs/blob/4aad6aa53df89513fb25d53a5805c55ea0b07693/python/deltalake/table.py#L54 (And sidenote: I just noticed the typehint on that configuration property is wrong; it should be `Dict[str, str]`, not `List[str]`. If you could correct that in this PR, it would be appreciated 👍 .)

File: python/tests/test_writer.py
Comment (fvaleye) [inline python/tests/test_writer.py:175]: **nitpick:** I would suggest renaming this method to `test_append_only_should_append_only_with_the_overwrite_mode` to improve the test readibility.

**Comment by fvaleye** [comment]
Thank you @WarSame 👍 

LGTM, ready to merge after fixing the [formatting issue](https://github.com/delta-io/delta-rs/runs/6123515763?check_suite_focus=true) in the CI.

**Comment by WarSame** [comment]
Sorry for the delay. I've had Covid for the last week so this has taken the backburner.

I believe I've fixed the formatting issues, though it appears the workflows need approval to run. I did add code to account for an existing table's metadata, and updated the tests to match that. Finally, it was getting to be a few lines of code to check these conditions so I pulled it out into its own functions. Please let me know if you prefer it inline.

File: python/deltalake/writer.py
Comment (wjones127) [inline python/deltalake/writer.py:209]: nit: I think this function name could be a bit shorter. What do you think of `__enforce_append_only()`? I think the existing name would work well as a doc string.

File: python/tests/test_writer.py
Comment (wjones127) [inline python/tests/test_writer.py:196]: I like that it checks the error message 👍

File: python/tests/test_writer.py
Comment (wjones127) [inline python/tests/test_writer.py:199]: If it errors on the first of these, we aren't testing the second, right? Or if it doesn't error on the first line, the test could still pass but we wouldn't like that behavior!

So maybe better to make those separate iterations:

```python
# Test both passing in path string and DeltaTable
possible_args = [path, table]
for arg, mode in itertools.product(possible_args, fail_modes):
    with pytest.raises(...):
        write_deltalake(arg, sample_data, mode=mode)

```

File: python/tests/test_writer.py
Comment (wjones127) [inline python/tests/test_writer.py:202]: To be extra safe, you could also test that the `DeltaTable.version()` hasn't changed since the last expected append. (It should equal 1, I think.)

**Comment by wjones127** [comment]
> Sorry for the delay. I've had Covid for the last week so this has taken the backburner.

Oh no! Well I hope you are feeling better now.

> Finally, it was getting to be a few lines of code to check these conditions so I pulled it out into its own functions. Please let me know if you prefer it inline.

Yeah that looks good to me.

**Review by wjones127 — CHANGES_REQUESTED** [review]
I have a few suggestions for the test.

File: python/deltalake/writer.py
Comment (fvaleye) [inline python/deltalake/writer.py:213]: Could you please add a docstring to this method (you could also mention that it will throw an exception)?

File: python/deltalake/writer.py
Comment (fvaleye) [inline python/deltalake/writer.py:221]: **nitpick:** add the mode in the `ValueError` to get the current value.

File: python/tests/test_writer.py
Comment (WarSame) [reply python/tests/test_writer.py]: Thank you! I was not aware of this feature!

If it had errored on the first we wouldn't have tested the second, but if we hadn't errored on the first we should have tested the second. Either way, I had intended to test both, and your recommended method is great for that.

File: python/deltalake/writer.py
Comment (WarSame) [reply python/deltalake/writer.py:211]: Please let me know if you want more in the docstring.

File: python/deltalake/writer.py
Comment (WarSame) [reply python/deltalake/writer.py]: Thanks for the callout! This is a great point to help people who might cross this error.

**Comment by wjones127** [comment]
@WarSame Looks like you just need to sort the imports, and then this PR is good-to-go. 😄

**Comment by WarSame** [comment]
@wjones127 I've done so and it should be good to go!

**Review by wjones127 — APPROVED** [review]

**Pull request merged**
